server/src/main/java/io/druid/server/listener/resource/ListenerResource.java

```diff
@@ -0,0 +1,183 @@
+ * This is a simple announcement resource that handles simple items that have a POST to an announcement endpoint, a
+ * GET of something in that endpoint with an ID, and a DELETE to that endpoint with an ID.
+ *
+ * The idea of this resource is simply to have a simple endpoint for basic POJO handling assuming the POJO has an ID
+ * which distinguishes it from others of its kind.
+ *
+ * This resource is expected to NOT block for POSTs, and is instead expected to make a best effort at returning
+ * as quickly as possible. Thus, returning ACCEPTED instead of OK is normal for POST methods here.
+ *
+ * Items tagged with a particular ID for an announcement listener are updated by a POST to the announcement listener's
+ * path "/{announcement}"
+ *
+ * Discovery of who can listen to particular announcement keys is not part of this class and should be handled
+ * by ListenerResourceAnnouncer
+ */
+public abstract class ListenerResource
+{
+  public static final String BASE_PATH = "/druid/listen/v1";
+  private static final Logger log = new Logger(ListenerResource.class);
+
+  private final ObjectMapper jsonMapper;
+  private final ObjectMapper smileMapper;
+  private final ListenerHandler handler;
+
+  public ListenerResource(
+      final @Json ObjectMapper jsonMapper,
+      final @Smile ObjectMapper smileMapper,
+      final ListenerHandler handler
+  )
+  {
+    this.jsonMapper = Preconditions.checkNotNull(jsonMapper);
+    this.smileMapper = smileMapper;
+    this.handler = handler;
+  }
+
+  @POST
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response serviceAnnouncementPOSTAll(
+      final InputStream inputStream,
+      final @Context HttpServletRequest req // used only to get request content-type
+  )
+  {
+    final String reqContentType = req.getContentType();
+    final boolean isSmile = SmileMediaTypes.APPLICATION_JACKSON_SMILE.equals(reqContentType);
+    final ObjectMapper mapper = isSmile ? smileMapper : jsonMapper;
+    try {
+      return handler.handlePOSTAll(inputStream, mapper);
+    }
+    catch (Exception e) {
+      log.error(e, "Exception in handling POSTAll request");
+      return Response.serverError().entity(ServletResourceUtils.sanitizeException(e)).build();
+    }
+  }
+
+  @GET
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response getAll()
+  {
+    try {
+      return handler.handleGETAll();
+    }
+    catch (Exception e) {
+      log.error(e, "Exception in handling GETAll request");
+      return Response.serverError().entity(ServletResourceUtils.sanitizeException(e)).build();
+    }
+  }
+
+  @Path("/{id}")
+  @GET
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response serviceAnnouncementGET(
+      final @PathParam("id") String id
+  )
+  {
+    if (Strings.isNullOrEmpty(id)) {
```

Comment:
What do you mean? can you expand a bit more on what you are wanting in a function?